src/MyParcelComApi.php

```diff
@@ -439,42 +444,59 @@ public function updateShipment(ShipmentInterface $shipment)
     }
 
     /**
-     * Determine what service contract to use for given shipment and update the
-     * shipment.
+     * Determine which service and contract to use for given shipment
+     * and update the shipment.
      *
      * @param ShipmentInterface $shipment
      * @return $this
      */
-    protected function determineContract(ShipmentInterface $shipment)
+    protected function determineServiceAndContract(ShipmentInterface $shipment)
     {
-        $selector = new ContractSelector();
         $calculator = new PriceCalculator();
 
-        $allServices = [];
-        $collection = $this->getServices($shipment);
-        for ($offset = 0; $offset < $collection->count(); $offset += 30) {
-            $allServices = array_merge($allServices, $collection->offset($offset)->get());
+        if ($shipment->getService() !== null) {
+            $services = [$shipment->getService()];
+        } else {
+            $services = [];
+            $collection = $this->getServices($shipment);
+            for ($offset = 0; $offset < $collection->count(); $offset += 30) {
+                $services = array_merge($services, $collection->offset($offset)->get());
+            }
         }
 
-        $contracts = array_map(
-            function (ServiceInterface $service) use ($selector, $calculator, $shipment) {
-                $contract = $selector->selectCheapest($shipment, $service->getServiceContracts());
+        $serviceIds = implode(',', array_map(function (ServiceInterface $service) {
+            return $service->getId();
+        }, $services));
+
+        $filters = [
+            'weight'  => $shipment->getPhysicalProperties()->getWeight(),
+            'service' => $serviceIds,
+        ];
+
+        if ($shipment->getContract() !== null) {
+            $filters['contract'] = $shipment->getContract()->getId();
+        };
+
+        $serviceRates = $this->getServiceRates($filters)->get();
 
+        $rates = array_map(
+            function (ServiceRateInterface $serviceRate) use ($calculator, $shipment) {
                 return [
-                    'price'    => $calculator->calculate($shipment, $contract),
-                    'contract' => $contract,
-                    'service'  => $service,
+                    'price'    => $calculator->calculate($shipment, $serviceRate),
+                    'service'  => $serviceRate->getService(),
+                    'contract' => $serviceRate->getContract(),
                 ];
             },
-            $allServices
+            $serviceRates
```

Comment:
These will only hold the first set of service rates. Instead of using the `array_map()` method, you can just use a `for`-loop. The `Collection` implements the `Iterator` interface, and will let you loop over all items (not just te first set). So that would guarantee that you would iterate over all service rates.